### ISSUE TITLE
Much slower QpMatrixStuffing on 1.1.18 compared to 1.1.7

### ISSUE THREAD

**Issue by hailiangliu89** [cvxpy/cvxpy#1611]
We noticed the same problem can be compiled much faster on 1.1.7 than the latest 1.1.18, with the same solving time via CPLEX 12.10.

On 1.1.18, it takes more than two minutes to compile, while on 1.1.7 the number is less than 20 seconds.

We suspected if enabling `OpenMP `(introduced in 1.1.11) made a difference, so we run tests on it as well. It did not show any difference on 1.1.18, whether `num_threads `was set to 4, 1 or not set at all.

We also profiled the `.solve()` function and noticed it is `build_matrix()` that takes longer on the newer version.

On 1.1.18:
`   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        2  128.084   64.042  128.084   64.042 {built-in method _cvxcore.build_matrix}
        1   19.370   19.370   19.376   19.376 {built-in method cplex._internal.py37_cplex12100.CPXXmipopt}
  1034058    3.956    0.000    3.956    0.000 {method 'tolist' of 'numpy.ndarray' objects}
        1    2.604    2.604   30.400   30.400 cplex_qpif.py:73(solve_via_data)
        2    1.848    0.924    1.848    0.924 {built-in method cplex._internal.py37_cplex12100.CPXXaddrows}
        2    1.507    0.754    1.507    0.754 {built-in method scipy.sparse._sparsetools.csr_sort_indices}
        2    1.252    0.626    1.252    0.626 {built-in method cplex._internal.py37_cplex12100.Pylolmat_to_CHBmat}`

On 1.1.7:
`   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1   18.430   18.430   18.434   18.434 {built-in method cplex._internal.py37_cplex12100.CPXXmipopt}
        2   11.613    5.806   19.133    9.567 canonInterface.py:61(reduce_problem_data_tensor)
        2    8.357    4.178    8.357    4.178 {built-in method _cvxcore.build_matrix}
        2    6.041    3.020    6.041    3.020 canonInterface.py:111(<dictcomp>)
  1034058    3.930    0.000    3.930    0.000 {method 'tolist' of 'numpy.ndarray' objects}
        1    2.604    2.604   28.938   28.938 cplex_qpif.py:79(solve_via_data)
        2    1.932    0.966    1.932    0.966 {built-in method scipy.sparse._sparsetools.csr_sort_indices}
        2    1.847    0.923    1.847    0.923 {built-in method cplex._internal.py37_cplex12100.CPXXaddrows}
        2    1.226    0.613    1.226    0.613 {built-in method cplex._internal.py37_cplex12100.Pylolmat_to_CHBmat}`

On 1.1.18, it is possible to enable verbose logging, so we noticed it is QpMatrixStuffing that takes almost all the time spent on compilation:
`-------------------------------------------------------------------------------
                                  Compilation                                  
-------------------------------------------------------------------------------
(CVXPY) Jan 22 05:02:55 PM: Compiling problem (target solver=CPLEX).
(CVXPY) Jan 22 05:02:55 PM: Reduction chain: EvalParams -> CvxAttr2Constr -> Qp2SymbolicQp -> QpMatrixStuffing -> CPLEX
(CVXPY) Jan 22 05:02:55 PM: Applying reduction EvalParams
(CVXPY) Jan 22 05:02:55 PM: Applying reduction CvxAttr2Constr
(CVXPY) Jan 22 05:02:56 PM: Applying reduction Qp2SymbolicQp
(CVXPY) Jan 22 05:02:56 PM: Applying reduction QpMatrixStuffing
(CVXPY) Jan 22 05:05:10 PM: Applying reduction CPLEX
(CVXPY) Jan 22 05:05:11 PM: Finished problem compilation (took 1.352e+02 seconds).
`
We also attached the full logs here, including the profiling stats.
[logs_with_cvxpy1118_MP0.txt](https://github.com/cvxpy/cvxpy/files/7922373/logs_with_cvxpy1118_MP0.txt)

[logs_with_cvxpy117_MP0.txt](https://github.com/cvxpy/cvxpy/files/7922368/logs_with_cvxpy117_MP0.txt)
.

**Comment by rileyjmurray** [comment]
Thanks so much for this detailed information, @hailiangliu89! Would you be able to share the code that produces these problems? If you aren't comfortable with sharing it publicly then you can email me and the other project maintainers. If this is an industrial problem then I'd be happy to sign an NDA if need be. Improving the speed of cvxcore is important for us and we need to investigate performance regressions closely.

**Comment by hailiangliu89** [comment]
I just attached the Problem saved with shelve that can be sent to `.solve()` directly.

**Comment by SteveDiamond** [comment]
How did you install 1.1.18? I'm pretty sure you need to install from source to use OpenMP. @akshayka is that right?

**Comment by akshayka** [comment]
That's correct (https://github.com/cvxpy/cvxpy/releases/tag/v1.1.11). @hailiangliu89 has verified that OpenMP is not relevant to this issue, though. (OpenMP was introduced in 1.1.11).

**Comment by SteveDiamond** [comment]
Sure, but it explains why he's not seeing any speedup with OpenMP. @hailiangliu89 I tried to use shelf but it's unclear to me how to reconstruct the problem object from your data. Could you please post a script that solves the problem?

**Comment by hailiangliu89** [comment]
`
import shelve
with shelve.open('Problem.prob') as ps: # unzip first and all three files are needed
     prob_shelve = ps['Problem']
prob_shelve.solve(solver=cp.CPLEX, verbose=True)
`

**Comment by SteveDiamond** [comment]
Thanks! We are looking into your example as part of a general rewrite of the backend.

**Comment by LiaoSW** [comment]
any improvements now?

**Comment by SteveDiamond** [comment]
There have been. Could you try 1.4.1 and report back to us?
